### ISSUE TITLE
Add EiffelPipelineDefinedEvent

### ISSUE THREAD

**Issue by e-backmark-ericsson** [Ericsson/eiffel#183]
### Description
There is a need to inform the eco system of tools surrounding the SW pipeline about the contents of a coming/started pipeline. That information should probably mainly focus on activities expected to be executed within the pipeline. Such an EiffelPipelineDefinedEvent could either be stand-alone or connected to a SCC event, depending on if the pipeline definition is delivered as code within a source code repository or not.

### Motivation
It would enable simpler, more generic and more powerful tools surrounding the SW pipeline e.g. providing visualizations and statistics.

### Exemplification
The new event should contain information about the activities expected to be executed and how they relate to each other. This would for example enable a 'follow-my-commit' visualization to draw a graph of activities expected to be executed before they have actually been triggered. I haven't thought much about the format yet since I need to understand if there is an interest in this kind of event before I put too much effort into defining it.

### Benefits
With this additional event the tools surrounding a SW pipeline can be prepared for events expected to come in the future. Currently those tools can just look a actual events and make assumptions on coming events based on historically linked events. With this, there is no need to guess anymore.

### Possible Drawbacks
Since this is a new event, and I expect no current events to have mandatory links to this new event, there should be no drawbacks.

**Comment by d-stahl-ericsson** [comment]
A couple of quick reflections. First, it would need to be extremely clear what the purpose is. As a user I might interpret this as a recipe which will instruct downstream behavior, or I might interpret it as a set of expected consequences, which may then be ticked off and compared to actual outcomes. 

The second is that it would seem to require rather a complicated data structure, and still it would be difficult to capture all the possible variations and corner cases. I guess from where I'm standing it just looks a lot like a big can of worms :)

Nevertheless, it would be interesting to hear some more thoughts on this. I would recommend starting a thread in the Google Group, my experience is you will get more eyes on it that way.

**Comment by guwenqing** [comment]
We in Zipline today are using EiffelConfigurationNotificationEvent in the earlier version of Eiffel (not open source) to notify the surrounding tools about a flow. It can be much easier for these tools to work when a model is available (like a recipe as Daniel say).
I do agree the data structure may be very complicated, is it possible to start without a data structure before we are more clear about what is a good one.

**Comment by e-backmark-ericsson** [comment]
I see the need primarily being a set of expected consequences and their relations.

Let's continue this discussion on https://groups.google.com/forum/#!topic/eiffel-community/z4cGaOOtz1Q

**Comment by guwenqing** [comment]
We in Zipline today are using EiffelConfigurationNotificationEvent in the earlier version of Eiffel (not open source) to notify the surrounding tools about a flow. It can be much easier for these tools to work when a model is available (like a recipe as Daniel say).
I do agree the data structure of the model may be very complicated, and it is not very easy to agree on a single model as well. Probably a protocol should allow multiple models.
Is it possible to start with a blob?

**Comment by e-backmark-ericsson** [comment]
@guwenqing, could you post the syntax of the flow config in the CNE event you're using for this purpose today, either here or in the Google group? Maybe it needs to be reformed a bit to match the opensource protocol syntax before publishing, so that we can have a discussion with correct references.

**Comment by d-stahl-ericsson** [comment]
I don't see any purpose of defining an event type with undefined syntax. Unless we have a good idea of how to express the information in a canonical way, we're better off using custom event types. In fact, that's a good route towards standardization: experiment and devise a syntax in a custom event type, and then bring that into the protocol.

**Comment by guwenqing** [comment]
@e-backmark-ericsson it is far way too complex, itself looks like a protocol as well.
@d-stahl-ericsson I get your point, but I don't think here is the same thing. 
First of all, in Eiffel level, I think should allow different schemas of the flow exist, inside one company, of course should try to agree on one schema. Defining the schema is another level of standardisation. 
Second, the schema format is very complex, probably same or even more than the protocol. Incorporate that syntax as part of the Eiffel protocol, I am not sure if it is a good idea. 
Third, it is always good to have such an event, sometimes no need to spread the definition of the flow as an attachment of the event, a notification of certain flow with a reference id is useful enough.
Go for the custom event is only working in theory, it can cause a lot of problems in the receiver side, don't even know if Eiffel components themselves can handle it well in practice. Our team wants to start with that, and I paused them, instead would like to have a dialog here (joining Emil).

**Comment by d-stahl-ericsson** [comment]
@guwenqing There is clearly potential value in the information of such an event, but a poorly or loosely formed event type can easily cause more problems than it solves. It is much better to have a narrow protocol that captures a few concepts well and concisely, than one that tries to do too much and falls short. That is why it's better to err on the side of caution.

So we need to get our concepts straight first. If we can manage that, syntax tends to follow. One thing that is not entirely clear is whose expectations on downstream behavior the event would represent. In practice, pipelines tend to not really be pipelines, but paths through a graph, and one stakeholder's idea of the pipeline may be a small subset of another stakeholder's idea. Think of a common component that is passed through a component pipeline, then perhaps some other integration verification pipeline. And then, when it passes through that, it would be consumed by a large number of product pipelines, et cetera. Asking any one stakeholder along the way, you would get different views of what the pipeline is. A big part of Eiffel's raison d'être is that you typically don't know exactly what will happen throughout a source revision's entire lifecycle at the time you create it, so there has to be some very clear scoping here.

Another thing to consider is how much detail should be included. If we publish expectations on _what_ will happen downstream, then perhaps we should also allow for expectations on **how**. E.g. how much time is allowed to pass between certain points in the expected pipeline would enable raising alerts when thresholds are crossed. And if we're going to do that, then perhaps also expectations on which actions should be taken. And then we're at a point where we have a master event that controls and configures the entire pipeline. There's a gradient here from complete choreography to complete orchestration, and by introducing this event type we place ourselves somewhere along that gradient, whether we want to or not. So we need to be very sure where we want to be and why.

**Comment by guwenqing** [comment]
Yes I agree about the event definition and so on, it must be well defined before released.
The point I am arguing right now is "do we really need to be that in detail regarding the protocol definition" as part of the event itself? Or the event is actually a notification that certain pipeline is deployed, and it is up to the pipeline to give extra information if it suits them.

I agree, perhaps we need a thorough thinking about all such events in the whole deployment cycle, which we are quite a bit far away to. In Zipline today, we only notify when the deployment starts, but we actually have no control on if the deployment is successful or not. Not very easy to do that without a master node as well since the deployment happens in a distributed manor.

In another mail thread that I am trying to ask Eiffel architects group there are mentioning about continuous operation events which I think it could fit also.

**Comment by d-stahl-ericsson** [comment]
Yes, I think exploring the operations extension would be valuable here. Not only as it would hopefully address some of your needs, but also to provide validation of the extension. 

In the meantime, you might want to use [AnnP](https://github.com/Ericsson/eiffel/blob/master/eiffel-vocabulary/EiffelAnnouncementPublishedEvent.md) to experiment with notification syntax. It's basically a catch-all for making announcements, so you can put any syntax you like there. When/if something crystallizes from that we can always formalize that as a dedicated event type.

On a side note, I would like to see the use cases for generic types like AnnP gradually reduced over time. The operations extension is an example of that: it largely replaces what the AnnP documentation says it should be used for. Which is a good thing!
